include/deal.II/numerics/cell_data_transfer.h

```diff
@@ -0,0 +1,212 @@
+   *   container stores values that differ in size. A varying amount of data
+   *   may be packed per cell, if for example the underlying ValueType of
+   *   the VectorType container is a container itself.
+   * @param[in] coarsening_strategy Function deciding which data to store on
+   *   a cell whose children will get coarsened into.
+   */
+  CellDataTransfer(
+    const Triangulation<dim, spacedim> &                triangulation,
+    const std::function<value_type(
+      const std::vector<value_type> &children_indices)> coarsening_strategy =
+      &CoarseningStrategies<value_type>::check_equality);
+
+  /**
+   * Prepare the current object for coarsening and refinement.
+   *
+   * Stores the active_cell_indices of all active cells on the associated
+   * triangulation and attribute them to either persisting, refined or coarsened
+   * cells.
+   */
+  void
+  prepare_for_coarsening_and_refinement();
+
+  /**
+   * Transfer the information from the previous mesh to the updated one.
+   *
+   * Data from the previous mesh supplied by @p in will be transferred to the updated
+   * mesh and stored in @p out. @p out has to provide enough space to hold the
+   * transferred data, i.e. has to be of size `triangulation.n_active_cells()`.
+   */
+  void
+  unpack(const VectorType &in, VectorType &out);
+
+private:
+  /**
+   * Pointer to the triangulation to work with.
+   */
+  SmartPointer<const Triangulation<dim, spacedim>,
+               CellDataTransfer<dim, spacedim, VectorType>>
+    triangulation;
+
+  /**
+   * Function deciding on how to process data from children to be stored on the
+   * parent cell.
+   */
+  const std::function<value_type(
+    const std::vector<value_type> &children_indices)>
+    coarsening_strategy;
+
+  /**
+   * Container to temporarily store the iterator and active cell index
+   * of cells that persist.
+   */
+  std::map<const typename Triangulation<dim, spacedim>::cell_iterator,
+           const unsigned int>
+    persisting_cells_active_index;
+
+  /**
+   * Container to temporarily store the iterator and active cell index
+   * of cells that will be refined.
+   */
+  std::map<const typename Triangulation<dim, spacedim>::cell_iterator,
+           const unsigned int>
+    refined_cells_active_index;
+
+  /**
+   * Container to temporarily store the iterator of parent cells that will
+   * remain after coarsening along with the active cell indices of the
+   * corresponding children cells.
+   */
+  std::map<const typename Triangulation<dim, spacedim>::cell_iterator,
+           const std::set<unsigned int>>
+    coarsened_cells_active_index;
+
+#ifdef DEBUG
+  /**
+   * Number of active cells on the initial triangulation that has not been
+   * refined yet.
+   *
+   * It will be set in prepare_for_coarsening_and_refinement() and used to
+   * validate user inputs after refinement happened.
+   */
+  unsigned int n_active_cells_pre;
+#endif
```

Comment:
I think it would be better to just write
```cpp
(void)n_active_cells_pre;
```
instead of changing the class members in different modes.